File: services/api-aspnet-core/App/Controllers/AuthController.cs
using System.Security.Claims;
using System.Security.Cryptography;
using App.Auth;
using App.Data;
using App.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace App.Controllers;

[ApiController]
[Route("auth")]
public class AuthController(
    AppDbContext db,
    JwtService jwtService,
    IConfiguration configuration,
    IWebHostEnvironment env) : ControllerBase
{
    // GET /auth/login — redirect to GitHub OAuth
    [HttpGet("login")]
    public IActionResult Login()
    {
        var clientId = configuration["GITHUB_CLIENT_ID"] ?? "dev-client-id";
        var redirectUri = Uri.EscapeDataString(
            configuration["GITHUB_CALLBACK_URL"] ?? "http://localhost:8080/auth/callback");
        var githubUrl =
            $"https://github.com/login/oauth/authorize?client_id={clientId}&redirect_uri={redirectUri}&scope=user:email";
        return Redirect(githubUrl);
    }

    // GET /auth/callback — exchange code, upsert user, return JWT
    [HttpGet("callback")]
    public async Task<IActionResult> Callback([FromQuery] string code, CancellationToken ct)
    {
        if (string.IsNullOrWhiteSpace(code))
            return BadRequest(new { error = "Missing OAuth code." });

        var clientId = configuration["GITHUB_CLIENT_ID"] ?? string.Empty;
        var clientSecret = configuration["GITHUB_CLIENT_SECRET"] ?? string.Empty;

        using var http = new HttpClient();
        http.DefaultRequestHeaders.Add("Accept", "application/json");

        var tokenRes = await http.PostAsJsonAsync(
            "https://github.com/login/oauth/access_token",
            new { client_id = clientId, client_secret = clientSecret, code },
            ct);

        if (!tokenRes.IsSuccessStatusCode)
            return StatusCode(502, new { error = "GitHub token exchange failed." });

        var tokenBody = await tokenRes.Content.ReadFromJsonAsync<GitHubTokenResponse>(cancellationToken: ct);
        if (tokenBody?.AccessToken is null)
            return StatusCode(502, new { error = "No access token from GitHub." });

        http.DefaultRequestHeaders.Authorization =
            new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", tokenBody.AccessToken);
        http.DefaultRequestHeaders.Add("User-Agent", "pipeline-studio");

        var userRes = await http.GetAsync("https://api.github.com/user", ct);
        if (!userRes.IsSuccessStatusCode)
            return StatusCode(502, new { error = "GitHub user fetch failed." });

        var githubUser = await userRes.Content.ReadFromJsonAsync<GitHubUserResponse>(cancellationToken: ct);
        if (githubUser is null)
            return StatusCode(502, new { error = "Could not parse GitHub user." });

        var email = githubUser.Email ?? $"{githubUser.Login}@github.noemail";

        var user = await db.Users.FirstOrDefaultAsync(u => u.Email == email, ct);

        if (user is null)
        {
            user = new User
            {
                Email = email,
                Name = githubUser.Name ?? githubUser.Login ?? email,
                Provider = "github",
            };
            db.Users.Add(user);
        }
        else
        {
            user.Name = githubUser.Name ?? githubUser.Login ?? user.Name;
            user.UpdatedAt = DateTime.UtcNow;
        }

        await db.SaveChangesAsync(ct);

        var jwt = jwtService.GenerateToken(user.Id, user.Email, user.Name);
        return Ok(new { token = jwt });
    }

    // GET /auth/me — return current user
    [HttpGet("me")]
    [Authorize]
    public async Task<IActionResult> Me(CancellationToken ct)
    {
        var userId = GetUserId();
        if (userId is null) return Unauthorized();

        var user = await db.Users.AsNoTracking()
            .FirstOrDefaultAsync(u => u.Id == userId.Value, ct);

        if (user is null) return NotFound();

        return Ok(new
        {
            user.Id,
            user.Email,
            user.Name,
            user.Provider,
            user.CreatedAt,
            HasApiKey = user.ApiKey is not null,
        });
    }

    // POST /auth/logout — clear cookie, return Ok
    [HttpPost("logout")]
    public IActionResult Logout()
    {
        Response.Cookies.Delete("auth_token");
        return Ok(new { message = "Logged out." });
    }

    // POST /auth/api-key — generate api key
    [HttpPost("api-key")]
    [Authorize]
    public async Task<IActionResult> GenerateApiKey(CancellationToken ct)
    {
        var userId = GetUserId();
        if (userId is null) return Unauthorized();

        var user = await db.Users.FirstOrDefaultAsync(u => u.Id == userId.Value, ct);
        if (user is null) return NotFound();

        var key = Convert.ToBase64String(RandomNumberGenerator.GetBytes(32))
            .Replace("+", "-").Replace("/", "_").TrimEnd('=');

        user.ApiKey = key;
        user.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(ct);

        return Ok(new { apiKey = key });
    }

    // DELETE /auth/api-key — revoke api key
    [HttpDelete("api-key")]
    [Authorize]
    public async Task<IActionResult> RevokeApiKey(CancellationToken ct)
    {
        var userId = GetUserId();
        if (userId is null) return Unauthorized();

        var user = await db.Users.FirstOrDefaultAsync(u => u.Id == userId.Value, ct);
        if (user is null) return NotFound();

        user.ApiKey = null;
        user.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(ct);

        return Ok(new { message = "API key revoked." });
    }

    // POST /dev/token — development-only test JWT
    [HttpPost("/dev/token")]
    public IActionResult DevToken([FromBody] DevTokenRequest? req)
    {
        if (!env.IsDevelopment())
            return NotFound();

        var userId = req?.UserId ?? Guid.NewGuid();
        var email = req?.Email ?? "dev@example.com";
        var name = req?.Name ?? "Dev User";

        var token = jwtService.GenerateToken(userId, email, name);
        return Ok(new { token });
    }

    private Guid? GetUserId()
    {
        var raw = User.FindFirstValue(ClaimTypes.NameIdentifier)
               ?? User.FindFirstValue("sub");
        return Guid.TryParse(raw, out var id) ? id : null;
    }
}

// ---- DTOs ----

public record DevTokenRequest(Guid? UserId, string? Email, string? Name);

internal sealed record GitHubTokenResponse(
    [property: System.Text.Json.Serialization.JsonPropertyName("access_token")]
    string? AccessToken);

internal sealed record GitHubUserResponse(
    [property: System.Text.Json.Serialization.JsonPropertyName("login")]
    string? Login,
    [property: System.Text.Json.Serialization.JsonPropertyName("name")]
    string? Name,
    [property: System.Text.Json.Serialization.JsonPropertyName("email")]
    string? Email);

File: services/api-laravel/app/Http/Controllers/AuthController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Http\JsonResponse;
use Illuminate\Support\Facades\Http;
use Illuminate\Support\Str;
use App\Models\User;
use Firebase\JWT\JWT;

class AuthController extends Controller
{
    // GET /auth/login → redirect to GitHub OAuth
    public function login(): JsonResponse
    {
        $clientId = config('services.github.client_id');
        $redirect  = config('services.github.redirect');

        $url = 'https://github.com/login/oauth/authorize'
            . '?client_id=' . $clientId
            . '&redirect_uri=' . urlencode($redirect)
            . '&scope=user:email';

        return response()->json(['redirect_url' => $url]);
    }

    // GET /auth/callback → exchange code, upsert User, return JWT JSON
    public function callback(Request $request): JsonResponse
    {
        $code = $request->query('code');
        if (empty($code)) {
            return response()->json(['error' => 'Missing code'], 422);
        }

        $tokenResponse = Http::post('https://github.com/login/oauth/access_token', [
            'client_id'     => config('services.github.client_id'),
            'client_secret' => config('services.github.client_secret'),
            'code'          => $code,
            'redirect_uri'  => config('services.github.redirect'),
        ])->json();

        $accessToken = $tokenResponse['access_token'] ?? null;
        if (empty($accessToken)) {
            return response()->json(['error' => 'OAuth exchange failed'], 400);
        }

        $ghUser = Http::withToken($accessToken)
            ->get('https://api.github.com/user')
            ->json();

        $email = $ghUser['email'] ?? null;
        if (empty($email)) {
            $emails = Http::withToken($accessToken)
                ->get('https://api.github.com/user/emails')
                ->json();
            foreach ($emails as $e) {
                if ($e['primary'] ?? false) {
                    $email = $e['email'];
                    break;
                }
            }
        }

        if (empty($email)) {
            return response()->json(['error' => 'Could not retrieve email from GitHub'], 400);
        }

        $user = User::updateOrCreate(
            ['email' => $email],
            [
                'name'     => $ghUser['name'] ?? $ghUser['login'],
                'provider' => 'github',
            ]
        );

        return response()->json(['token' => $this->makeJwt($user), 'user' => $user]);
    }

    // GET /auth/me → auth required → return user
    public function me(): JsonResponse
    {
        return response()->json(auth()->user());
    }

    // POST /auth/logout → 200
    public function logout(): JsonResponse
    {
        return response()->json(['message' => 'Logged out']);
    }

    // POST /auth/api-key → generate and save API key
    public function generateApiKey(): JsonResponse
    {
        $user = auth()->user();
        $key  = Str::random(64);
        $user->update(['api_key' => $key]);

        return response()->json(['api_key' => $key]);
    }

    // DELETE /auth/api-key → revoke API key
    public function revokeApiKey(): JsonResponse
    {
        auth()->user()->update(['api_key' => null]);

        return response()->json(['message' => 'API key revoked']);
    }

    // POST /dev/token → local env only
    public function devToken(Request $request): JsonResponse
    {
        if (! app()->environment('local')) {
            return response()->json(['error' => 'Not available'], 403);
        }

        $email = $request->input('email', 'dev@example.com');
        $user  = User::firstOrCreate(
            ['email' => $email],
            ['name' => 'Dev User', 'provider' => 'local']
        );

        return response()->json(['token' => $this->makeJwt($user), 'user' => $user]);
    }

    private function makeJwt(User $user): string
    {
        $secret = config('app.jwt_secret');
        $now    = time();

        $payload = [
            'iss' => config('app.url'),
            'sub' => $user->id,
            'email' => $user->email,
            'name'  => $user->name,
            'iat'   => $now,
            // Session length is set by the active industry profile (HIPAA → 15 min).
            'exp'   => $now + \App\Support\Compliance::active()->sessionTimeoutSeconds,
        ];

        return JWT::encode($payload, $secret, 'HS256');
    }
}
